lighthouse-core/scripts/extract-crdp-mapping.js

```diff
@@ -147,39 +178,205 @@ function getEventMap(sourceRoot, domainName) {
-    // Don't append type path name to void event payload types
-    const eventType = rawEventType === 'void' ? 'void' :
-        `Crdp.${domainName}.${rawEventType}`;
+    let eventType = 'void';
+    if (rawEventType !== null) {
+      const {domain, type} = getTypeName(rawEventType, methodName);
+      eventType = `Crdp.${domain}.${type}`;
+    }
 
     eventMap.set(eventName, eventType);
   }
 
   return eventMap;
 }
 
+/**
+ * Asserts that this is a function returning a promised value and returns the
+ * type of that value.
+ * @param {ts.FunctionTypeNode} functionNode
+ * @param {string} debugName A name to print for function if error is found.
+ * @return {string}
+ */
+function getPromisedReturnType(functionNode, debugName) {
+  const returnTypeNode = functionNode.type;
+  if (returnTypeNode && ts.isTypeReferenceNode(returnTypeNode)) {
+    // Check returning a promise.
+    if (!ts.isIdentifier(returnTypeNode.typeName) ||
+        returnTypeNode.typeName.text !== 'Promise' ||
+        !returnTypeNode.typeArguments) {
+      throw new Error(`command ${debugName} has unexpected return type`);
+    }
+
+    // Get promise's payload
+    if (returnTypeNode.typeArguments.length !== 1) {
+      throw new Error(`unexpected param(s) passed to ${debugName}`);
+    }
+
+    const payloadType = returnTypeNode.typeArguments[0];
+
+    if (payloadType.kind === ts.SyntaxKind.VoidKeyword) {
+      return 'void';
+    } else if (ts.isTypeReferenceNode(payloadType)) {
+      const {domain, type} = getTypeName(payloadType, debugName);
+      return `${domain}.${type}`;
+    }
+  }
+
+  throw new Error(`unexpected return type for ${debugName}`);
+}
+
+/**
+ * Returns true if all properties on interface are optional. Does simple search
+ * for interface declaration, assuming top level domain name and interface
+ * defined at the top level within.
+ * @param {ts.Node} sourceRoot
+ * @param {string} domainName
+ * @param {string} interfaceName
+ * @return {boolean}
+ */
+function isWeakInterface(sourceRoot, domainName, interfaceName) {
+  const domainInterface = findFirstInterfaceOrModule(sourceRoot, domainName);
+  if (!domainInterface || !ts.isModuleDeclaration(domainInterface)) {
+    throw new Error(`domain ${domainName} not found`);
+  }
+  const targetInterface = findFirstInterfaceOrModule(domainInterface, interfaceName);
+  if (!targetInterface || !ts.isInterfaceDeclaration(targetInterface)) {
+    throw new Error(`interface ${interfaceName} not found within ${domainName} domain`);
+  }
+
+  return targetInterface.members.every(member => {
+    if (!ts.isPropertySignature(member)) {
+      return true;
+    }
+
+    return member.questionToken !== undefined;
+  });
+}
+
+/**
+ * Returns a Map of events to params and return type for given domain.
+ * @param {ts.Node} sourceRoot
+ * @param {string} domainName
+ * @return {Map<string, {paramsType: string, returnType: string}>}
+ */
+function getCommandMap(sourceRoot, domainName) {
```

Comment:
do you want to upstream all this into `devtools-protocol`?

Reply:
> do you want to upstream all this into devtools-protocol?

the event map (just `'Domain.eventName': Crdp.EventPayloadType` I think could easily live upstream, it's so simple. The command map seems much more specialized to our use case, so I'm not sure. We could propose it to everyone involved in https://github.com/ChromeDevTools/devtools-protocol/pull/90 and see what they think

Reply:
cc @@JoelEinbinder whatchu think homeboy

Reply:
he sounded 👍 for upstreaming (if not for using in pptr)